src/service-worker/service-worker.ts

```diff
@@ -1,37 +1,57 @@
-import {Message, FileRecord} from '../shared/types.js';
-import {endWithSlash} from '../shared/util.js';
-import {expose} from 'comlink';
-import {MESSAGE_TYPES} from '../shared/constants.js';
-
-const swScope = (self as unknown) as ServiceWorkerGlobalScope;
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  HANDSHAKE_RECEIVED,
+  ESTABLISH_HANDSHAKE,
+  ServiceWorkerAPI,
+  FileAPI,
+} from '../shared/worker-api.js';
+import { endWithSlash } from '../shared/util.js';
+import { expose } from 'comlink';
+
+declare var self: ServiceWorkerGlobalScope;
+
+/**
+ * API exposed to the UI thread via Comlink. The static methods on this class
+ * become instance methods on SwControllerAPI.
+ */
+const workerAPI: ServiceWorkerAPI = {
+  setFileAPI(fileAPI: FileAPI, sessionID: string) {
+    fileAPIs.set(sessionID, fileAPI);
```

Comment:
Ugh this is so clean I'm jealous

Reply:
It's kinda amazing how simple and powerful SWs are yet such a PiTA to use when it comes to the nitty semantics

Reply:
Thanks! And ell me about it. I just got the last flakiness issue ironed out too I think.